Add setting to match completion matches case insensitive

grimm26 opened the pull request.

grimm26 committed:
Add setting to match completion matches case sensitive

details in the README

grimm26 commented:
After I submitted this, I realize that it supersedes https://github.com/zimfw/completion/pull/2 which seems to be abandoned.

ericbn commented on init.zsh:
```diff
@@ -42,7 +47,11 @@
   zstyle ':completion:*' verbose yes
-  zstyle ':completion:*' matcher-list 'm:{a-zA-Z}={A-Za-z}' '+r:|?=**'
+  if zstyle -t ':zim:completion' case-sensitive; then
+    zstyle ':completion:*' matcher-list '' 'm:{a-zA-Z}={A-Za-z}' '+r:|?=**'
```
What about just having:
```suggestion
    zstyle ':completion:*' matcher-list 'r:|?=**'
```
here?

grimm26 replied:
oof, not what I wanted. With default case insensitive setup, it only shows matches for the middle of a word if none match anchored to the beginning. Your suggestion shows middle and beginning matches always.

ericbn commented on init.zsh:
```diff
@@ -17,8 +17,13 @@
 
-  # Make globbing case insensitive.
-  setopt NO_CASE_GLOB
+  if zstyle -t ':zim:completion' case-sensitive; then
```
I've been considering having this option, and I think:
* it should have a more global name (not module-specific), since many modules can use the same option to configure themselves regarding case sensitivity.
* value should allow for more than `yes`/`no` as option, so it could allow for `sensitive`, `insensitive` and `smart`, for example.

grimm26 replied:
smart being to be insensitive if the provided case is not present? Sounds like a more comprehensive solution. What to call the option...

grimm26 replied:
Start a 'zim:global` namespace or something? 

`zstyle 'zim:global' case-sensitivity smart`

grimm26 committed:
Set up a ':zim:case:*' style that can be insensitive, sensitive, or
smart.

- insensitive is the default and how zimfw has worked up until now.
- sensitive flips NO_CASE_GLOB to CASE_GLOB and eliminates the completion
matcher-list that makes completion matches case insensitive.
- smart also sets CASE_GLOB but instead modifies the matcher-list to first
match the supplied case and tgen try being case insensitive if no
initial match was found.

I made it ':zim:case:*' so that more specific styles could be put into
place if people wanted case sensitive for some things and insensitive
for others.

grimm26 commented:
@ericbn How's this?

ericbn committed:
Update README.md and init.zsh

ericbn commented:
Hi @grimm26. Just updated the PR with my suggestion. I'm simplifying the zstyle to be `zstyle ':zim:*' case-sensitivity sensitive` and allowing just configuring between sensitive/insensitive for now. The smart option is just something for a probable future. I like the idea of allowing configuring each feature separately and then using `*` for both. That's exactly how zstyle is supposed to be used!  :- )

grimm26 committed:
Add smart case-sensitivity for completions

grimm26 commented:
"smart" is my preferred setting for matcher-list, so I added that back with a slight tweak from what I originally had.

ericbn commented:
Hi @grimm26. The general definition of "smart" is: "Searches case insensitively if the pattern is all lowercase. Search case sensitively otherwise." Not sure how to implement that with matcher-list in a straightforward way. This match pattern gets close to that: `'m:{a-z}={A-Z}'`. It will actually match lower case letters insensitively and upper case letters as is.

For example:
```
❯ touch README.md Readme.tex readme.txt
```
Case insensitive:
```
❯ vim re<TAB>
-- file --
README.md   Readme.tex  readme.txt

❯ vim Re<TAB>
-- file --
README.md   Readme.tex  readme.txt
```
Case sensitive:
```
❯ vim re<TAB>
❯ vim readme.txt

❯ vim Re<TAB>
❯ vim Readme.tex
```
Expected smart:
```
❯ vim re<TAB>
-- file --
README.md   Readme.tex  readme.txt

❯ vim Re<TAB>
❯ vim Readme.tex

❯ vim RE<TAB>
❯ vim README.md
```
Above "half-smart":
```
❯ vim re<TAB>
-- file --
README.md   Readme.tex  readme.txt

❯ vim Re<TAB>
-- file --
README.md   Readme.tex

❯ vim RE<TAB>
❯ vim README.md
```

grimm26 commented:
OK, perhaps I am mis-using "smart" then. I guess what I have is more "fallback" case insensitivity :). I can back out my "smart" setting and I'll just manually set the matcher-set how I want it in my zshrc. Basically, i can't stand when I tab complete "re" and it suggests README instead of just giving me "react/" or whatever.

grimm26 commented:
@ericbn Also, is there a preference on this project for me to squash the commits in this PR before we merge it to avoid spam of multiple incremental commits in master?

grimm26 committed:
Revert "Add smart case-sensitivity for completions"

This reverts commit 1c032d77e35239db78a8822c27d1d0e304f67f42.

ericbn committed:
Update init.zsh

ericbn commented:
Just squashed it when merging. Thanks for the contribution!

This part of the completion system is maybe the most complex part of the whole Zim settings, and the one that still does not work fully as expected. See https://github.com/zimfw/zimfw/issues/212. So kind of a sensitive piece...  :- )

grimm26 commented:
@ericbn Thanks. Yeah, there are still pieces that confuse me :) https://github.com/zimfw/zimfw/discussions/491
